lambeq/rewrite/rewrite_diagram.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2021-2023 Cambridge Quantum Computing Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Rewrite
+=======
+A rewrite rule is a schema for transforming/simplifying a diagram.
+A diagram rewriter is a function for transforming/simplifying a diagram at diagram level.
+
+The :py:class:`DiagramRewriter` applies the transformation based on the defined rules.
+
+Subclass :py:class:'DiagramRewriter' to define a custom rewrite rule. An
+example rewrite rule :py:class:`MergeWiresRewriter` has been provided for
+merging the free wires of a diagram with more than one free wire.
+"""
+from __future__ import annotations
+
+__all__ = ['DiagramRewriter',
+           'MergeWiresRewriter']
+
+from abc import ABC, abstractmethod
+from collections.abc import Container, Iterable
+
+from discopy import Word
+from discopy.rigid import Box, Cap, Cup, Diagram, Functor, Id, Spider, Swap, Ty
+from discopy.rigid import caps, spiders
+
+from lambeq.core.types import AtomicType
+
+N = AtomicType.NOUN
+S = AtomicType.SENTENCE
+
+
+class DiagramRewriter(ABC):
+    """Base class for diagram level rewrite rules."""
+
+    @abstractmethod
+    def matches(self, diagram: Diagram) -> bool:
+        """Check if the given diagram should be rewritten."""
+
+    @abstractmethod
+    def rewrite_diagram(self, diagram: Diagram) -> Diagram:
```

Comment:
We would also need a method that applies to a list of diagrams, e.g. `rewrite_diagrams`.

Reply:
I was wondering, what about a single method that given the diagram type being Diagram or list of Diagram, it would automatically perform the rewriting correctly and returning either a diagram or a list of diagrams.

Reply:
My understanding is that we rely on the __call__ method correct? No one should run rewrite_diagram because then it wouldn't check for matches.

Reply:
Regarding your second comment, from looking at the code It isnt clear to me whether `DiagramRewriter` is meant to be analogous to `RewriteRule` or `Rewriter`. 
In the current rewriting interface, the user does not call the `RewriteRule` on a box directly. This is the responsibility of the `Rewriter`, which is initialised with a list of `RewriteRule`s, and then when called on a diagram, applies these rules to it.
It might make sense to have a similar structure for the whole-diagram rewrites

Reply:
Exactly, it's supposed to be a merged version of RewriteRule and Rewriter since we do it at diagram level. I was initially going to do it with a init and base format, but was requested to instead to it in a single file, so I followed the current structure. 

I actually agree that it should be a bit more similar to the current rewrite rule and rewriter instance, but maybe that's not possible given we are performing the rewriting at diagram level? Functorially I mean.

Reply:
I don't think it makes much sense to have a single rewriter with many rules at the diagram level, similar to what we do at the box level. Box rules are local, but diagram "rules" are global to the diagram, so it doesn't make sense to apply more than one of them. And mixing box rules with diagram rules is not a good idea since the scope is so different.